File: packages/l402-ai-sdk/src/tools.ts
/**
 * Vercel AI SDK tool definitions for L402 payments.
 *
 * All L402 protocol logic runs in Rust via WASM — no TypeScript
 * reimplementation. This module provides thin tool wrappers around
 * the `WasmL402Client` from `l402-wasm`.
 *
 * @example
 * ```typescript
 * import { createL402Tools, WasmL402Client, WasmBudgetConfig } from '@lightninglabs/l402-ai';
 * import { generateText } from 'ai';
 * import { openai } from '@ai-sdk/openai';
 *
 * const client = WasmL402Client.withLndRest(
 *   'https://localhost:8080',
 *   process.env.LND_MACAROON!,
 *   new WasmBudgetConfig(1000, 0, 50000, 0),
 *   100,
 * );
 *
 * const tools = createL402Tools({ client });
 *
 * const result = await generateText({
 *   model: openai('gpt-4o'),
 *   tools,
 *   prompt: 'Fetch data from https://api.example.com/paid-endpoint',
 * });
 * ```
 */

import { tool } from 'ai';
import { z } from 'zod';

import type { WasmL402Client } from '@lightninglabs/l402';

/** Configuration for creating L402sdk AI SDK tools. */
export interface L402ToolsConfig {
  /** A `WasmL402Client` instance from `l402-wasm`. */
  client: WasmL402Client;
}

/**
 * Create Vercel AI SDK tools for L402 Lightning payments.
 *
 * Returns an object with tools that can be spread into the `tools` parameter
 * of `generateText` or `streamText`.
 *
 * Tools provided:
 * - `l402_fetch`: Fetch a URL, automatically paying L402 challenges
 * - `l402_get_receipts`: Get payment receipts for cost tracking
 */
export function createL402Tools(config: L402ToolsConfig) {
  const { client } = config;

  return {
    l402_fetch: tool({
      description:
        'Fetch a URL, automatically paying Lightning invoices for L402-gated APIs. ' +
        'When the server requires payment (HTTP 402), this tool pays the Lightning invoice, ' +
        'caches the token, and retries the request. Returns the response body and payment ' +
        'receipt if a payment was made. Use this for any API that requires L402 authentication.',
      inputSchema: z.object({
        url: z.string().url().describe('The URL to fetch'),
        method: z
          .enum(['GET', 'POST'])
          .default('GET')
          .describe('HTTP method to use'),
        body: z
          .string()
          .optional()
          .describe('Request body (for POST requests). Should be JSON-encoded.'),
      }),
      execute: async ({ url, method, body }) => {
        try {
          const response =
            method === 'POST'
              ? await client.post(url, body ?? undefined)
              : await client.get(url);

          const receipt = response.receipt;

          return {
            status: response.status,
            body: response.body,
            paid: response.paid,
            receipt: receipt
              ? {
                  amountSats: Number(receipt.amountSats),
                  feeSats: Number(receipt.feeSats),
                  totalCostSats: Number(receipt.totalCostSats()),
                  paymentHash: receipt.paymentHash,
                }
              : null,
          };
        } catch (e) {
          return {
            status: 0,
            body: '',
            paid: false,
            receipt: null,
            error: e instanceof Error ? e.message : String(e),
          };
        }
      },
    }),

    l402_get_receipts: tool({
      description:
        'Get all L402 payment receipts from this session. ' +
        'Useful for tracking costs, auditing payments, and reporting spend to the user.',
      inputSchema: z.object({}),
      execute: async () => {
        try {
          const totalSpent = await client.totalSpent();
          const receipts = await client.receipts();

          return {
            totalSpentSats: Number(totalSpent),
            paymentCount: Array.isArray(receipts) ? receipts.length : 0,
            receipts: Array.isArray(receipts)
              ? receipts.map((r: any) => ({
                  endpoint: r.endpoint,
                  amountSats: Number(r.amountSats),
                  feeSats: Number(r.feeSats),
                  totalCostSats: Number(r.totalCostSats()),
                  responseStatus: r.responseStatus,
                  timestamp: Number(r.timestamp),
                }))
              : [],
          };
        } catch (e) {
          return {
            totalSpentSats: 0,
            paymentCount: 0,
            receipts: [],
            error: e instanceof Error ? e.message : String(e),
          };
        }
      },
    }),
  };
}

File: packages/l402-langchain/src/l402_langchain/tools.py
"""LangChain tools for L402-gated API access via L402sdk.

Provides ``L402FetchTool`` for making HTTP requests that automatically
handle L402 payment challenges, and ``L402BudgetTool`` for monitoring
spending across all L402 interactions.
"""

from __future__ import annotations

import json
from typing import Any, ClassVar, Optional

from langchain_core.tools import BaseTool
from pydantic import ConfigDict, Field

from l402 import L402Client, L402Response


class L402FetchTool(BaseTool):
    """LangChain tool for fetching data from L402-gated APIs.

    Wraps a ``L402sdk.L402Client`` to handle the full L402 payment flow
    transparently: HTTP 402 -> parse challenge -> pay Lightning invoice ->
    retry with token -> return data.

    Budget limits configured on the client are enforced automatically.
    Token caching is handled by L402sdk so repeated requests to the same
    endpoint reuse cached tokens without additional payments.

    Example::

        from l402_langchain import create_l402_client, L402FetchTool

        client = create_l402_client(
            backend="lnd",
            url="https://localhost:8080",
            macaroon="deadbeef...",
        )
        tool = L402FetchTool(client=client)
        result = tool.invoke("https://api.example.com/data")
    """

    name: ClassVar[str] = "l402_fetch"
    description: ClassVar[str] = (
        "Fetch data from an API that requires Lightning payment (L402). "
        "Input should be a full URL, or a JSON object with 'url' and "
        "optional 'body' for POST requests. The tool automatically handles "
        "payment if the API responds with HTTP 402. Returns the response "
        "body with payment metadata."
    )

    model_config = ConfigDict(arbitrary_types_allowed=True)

    client: L402Client = Field(exclude=True)

    def _run(self, input_str: str) -> str:
        """Execute an L402-aware HTTP request.

        Args:
            input_str: Either a plain URL for GET requests, or a JSON
                string ``{"url": "...", "body": "..."}`` for POST requests.

        Returns:
            Response body prefixed with payment metadata. On error,
            returns a descriptive error string (not an exception) so the
            LLM agent can reason about the failure.
        """
        url, body = _parse_input(input_str)

        try:
            if body is not None:
                response: L402Response = self.client.post(url, body)
            else:
                response = self.client.get(url)
        except ValueError as exc:
            return f"Payment error: {exc}"
        except RuntimeError as exc:
            return f"Request error: {exc}"

        return _format_response(response)


class L402BudgetTool(BaseTool):
    """LangChain tool for monitoring L402 spending.

    Reports total satoshis spent, number of payments, and a per-endpoint
    breakdown. Useful for cost-aware agents that need to track or report
    their API spending.

    Example::

        from l402_langchain import create_l402_client, L402BudgetTool

        client = create_l402_client(
            backend="lnd",
            url="https://localhost:8080",
            macaroon="deadbeef...",
        )
        tool = L402BudgetTool(client=client)
        result = tool.invoke("")  # "No payments made yet. ..."
    """

    name: ClassVar[str] = "l402_check_budget"
    description: ClassVar[str] = (
        "Check total Lightning sats spent so far across all L402 API "
        "calls. Takes no input. Returns total sats spent, receipt count, "
        "and a breakdown per endpoint."
    )

    model_config = ConfigDict(arbitrary_types_allowed=True)

    client: L402Client = Field(exclude=True)

    def _run(self, _input: str = "") -> str:
        """Return a spending summary.

        Returns:
            Human-readable spending report with total sats, receipt
            count, and per-endpoint breakdown.
        """
        total = self.client.total_spent()
        receipts = self.client.receipts()

        if not receipts:
            return "No payments made yet. Total spent: 0 sats."

        lines = [
            f"Total spent: {total} sats across {len(receipts)} payment(s).",
            "",
        ]
        for i, receipt in enumerate(receipts, 1):
            lines.append(
                f"  #{i}: {receipt.endpoint} — "
                f"{receipt.amount_sats} sats "
                f"(+{receipt.fee_sats} fee, "
                f"status {receipt.response_status}, "
                f"{receipt.latency_ms}ms)"
            )

        return "\n".join(lines)


def _parse_input(input_str: str) -> tuple[str, Optional[str]]:
    """Parse tool input into (url, optional_body).

    Accepts either a plain URL string or a JSON object with 'url' and
    optional 'body' keys.
    """
    stripped = input_str.strip()

    # Try JSON parse for POST requests
    if stripped.startswith("{"):
        try:
            data = json.loads(stripped)
            url = data.get("url", "")
            body = data.get("body")
            if isinstance(body, dict):
                body = json.dumps(body)
            return url, body
        except (json.JSONDecodeError, AttributeError):
            pass

    # Plain URL for GET requests
    return stripped, None


def _format_response(response: L402Response) -> str:
    """Format an L402Response into a human-readable string with metadata."""
    if response.paid and response.receipt is not None:
        receipt = response.receipt
        header = (
            f"[Paid {receipt.amount_sats} sats "
            f"(+{receipt.fee_sats} fee) | "
            f"status {response.status}]\n"
        )
        return header + response.text()

    return f"[No payment | status {response.status}]\n{response.text()}"
